### PR TITLE
Micro-optimizations for Decompressor and Context

### PR DESCRIPTION
**Pull request by maruth-stripe**
We found `current_table_size`, `Decompressor#read_header` and `Context#decode` to be expensive functions while profiling code that used `protocol-http2`. Microbenchmarks show a ~15+% improvement in decoding performance.
## Types of Changes

- Performance improvement.

## Contribution
We unrolled the loop in `read_header` to a `case-when` block. Specifically using constants for the `when` clauses so Ruby uses a hash lookup. We also remove some unnecessary hash lookups. We also do bookkeeping for the current table size instead of recomputing.

The changes passed all tests locally. For benchmarking we used this slightly modified version of the rfc7541 test case
```ruby
require 'protocol/hpack/decompressor'
require 'yaml'


def fixtures(mode)
	Dir.glob(File.expand_path("rfc7541/*.yaml", __dir__)) do |path|
		fixture = YAML::load_file(path)
		
		if only = fixture[:only]
			next unless only == mode
		end
		
		yield fixture
	end
end

def timed(tag=nil)
	t = Time.now
	ret = yield
	d = Time.now - t
	d
end

t = 0
fixtures(:decompressor) do |example|
	example[:streams].size.times do |nth|
		t += timed do
			10_000.times do
				context = Protocol::HPACK::Context.new(huffman: example[:huffman], table_size: example[:table_size])
				buffer = String.new.b
				decompressor = Protocol::HPACK::Decompressor.new(buffer, context)
				(0...nth).each do |i|
					buffer << [example[:streams][i][:wire].gsub(/\s/, '')].pack('H*')
					decompressor.decode
				end
			end
		end
	end
end

puts t
```
- [X] I tested my changes locally.
- [X] I agree to the [Developer's Certificate of Origin 1.1](https://developercertificate.org/).

cc @froydnj

### CODE REVIEW

**Comment by ioquatix** [comment]
Wow, amazing!!

I've rebased your PR on top of a working CI, which includes external tests from `protocol-http2` and `async-http` so we have a fair idea that this doesn't break real usage.

I'd love for you to add benchmarks to `benchmarks/decompressor/...` with an included `readme.md` which includes details of how to run the benchmark and any relevant details/analysis/results/conclusions. It's useful for future work. I'm happy for you to do this in a separate PR.

For benchmark timing, I recommend you avoid `Time.now` as this can be fairly imprecise, and instead consider using the `Process.clock_gettime(Process::CLOCK_MONOTONIC)` or similar.

File: lib/protocol/hpack/context.rb [hunk 1]
```diff
@@ -12,6 +12,11 @@ module Protocol
 	# - http://tools.ietf.org/html/draft-ietf-httpbis-header-compression-10
 	module HPACK
 		# Header representation as defined by the spec.
+		NO_INDEX_TYPE = {prefix: 4, pattern: 0x00}.freeze
+		NEVER_INDEXED_TYPE = {prefix: 4, pattern: 0x10}.freeze
+		CHANGE_TABLE_SIZE_TYPE = {prefix: 5, pattern: 0x20}.freeze
+		INCREMENTAL_TYPE = {prefix: 6, pattern: 0x40}.freeze
+		INDEXED_TYPE = {prefix: 7, pattern: 0x80}.freeze
 		HEADER_REPRESENTATION = {
 			indexed: {prefix: 7, pattern: 0x80},
```
Comment (ioquatix) [inline lib/protocol/hpack/context.rb]: Would it make sense to use the constants defined above, here?

e.g.

```suggestion
			indexed: INDEXED_TYPE,
```

?

File: lib/protocol/hpack/context.rb [hunk 2]
```diff
@@ -296,6 +301,11 @@ def add_to_table(command)
 				command.freeze
 				
 				@table.unshift(command)
+				@cursize += entry_size(command)
```
Comment (ioquatix) [inline lib/protocol/hpack/context.rb]: Small nit, if we are going to introduce this, can we rename it to `@current_table_size`? I'm okay to do this in a separate PR if you want to keep this PR hygienic.

File: lib/protocol/hpack/decompressor.rb [hunk 3]
```diff
@@ -98,30 +101,56 @@ def read_header
...
-				header[:type], type = HEADER_REPRESENTATION.find do |_t, desc|
-					mask = (pattern >> desc[:prefix]) << desc[:prefix]
-					mask == desc[:pattern]
-				end
 
-				raise CompressionError unless header[:type]
+				type = nil
+
+
+				case (pattern & MASK_SHIFT_4)
```
Comment (ioquatix) [inline lib/protocol/hpack/decompressor.rb:134]: Can you add a brief comment explaining how the lookup here works, and where the values are coming from?

File: lib/protocol/hpack/decompressor.rb [hunk 4]
```diff
@@ -98,30 +101,56 @@ def read_header
...
+				# possible combinations of 2 bits (5th, 6th)
+				when 0x40, 0x50, 0x60, 0x70
+					header[:type] = :incremental
+					type = INCREMENTAL_TYPE
+				# checking if (pattern >> 7) << 7 == 0x80
+				when 0x80, 0x90, 0xa0, 0xb0, 0xc0, 0xd0, 0xe0, 0xf0
+					header[:type] = :indexed
+					type = INDEXED_TYPE
+				else
+					raise CompressionError
```
Comment (ioquatix) [inline lib/protocol/hpack/decompressor.rb]: We are missing coverage for this line - can you add a test for the failure case?

**Review by ioquatix — CHANGES_REQUESTED** [review]

File: lib/protocol/hpack/decompressor.rb (hunk 4 above)
Comment (maruth-stripe) [reply lib/protocol/hpack/decompressor.rb]: Never mind, this failure case is impossible (see comment)

**Comment by maruth-stripe** [comment]
Addressed comments, I'll add a benchmarking script in a follow-up!

**Review by ioquatix — APPROVED** [review]
Great, fantastic work.
